Implemented `probability` for discrete random variables

akash9712 opened the pull request:
This PR contains a basic implementation of the `probability` method for
`SingleDiscretePSpace`. Prior to this, probability for a given random
variable could be calculated for countinuous and finite random variables,
but a `NotImplementedError` was raised for the discrete case.
Tests were added.

Examples
```
>>> from sympy import *
>>> from sympy.stats import *
>>> X, Y = Geometric('X', S(1)/2), Poisson('Y', 4)
>>> P(X < 5)
15/16
>>> P(Y > 2)
32*(-39/32 + 3*exp(4)/32)*exp(-4)/3
```

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed


#### Other comments

akash9712 committed:
Implemented `probability` for discrete random variables

This PR contains a basic implementation of the `probability` method for
`SingleDiscretePSpace`. Prior to this, probability for a given random
variable could be calculated for countinuous and finite random variables,
but a `NotImplementedError` was raised for the discrete case.
Tests were added.

Examples
```
>>> from sympy import *
>>> from sympy.stats import *
>>> X, Y = Geometric('X', S(1)/2), Poisson('Y', 4)
>>> P(X < 5)
15/16
>>> P(Y > 2)
32*(-39/32 + 3*exp(4)/32)*exp(-4)/3
```

akash9712 committed:
Added newline at the end of drv.py

normalhuman commented on sympy/stats/tests/test_discrete_rv.py:
```diff
@@ -34,3 +36,12 @@ def test_sample():
+    assert P(X < 3) == S(9)/25
+    assert P(X > 3) == S(64)/125
+    assert P(Eq(Y, 3)) == 32*exp(-4)/3
+    assert P(Y < 3) == 13*exp(-4)
```
It's a fault of str printer, not yours. The expression that `P(Y > 3)` actually returns can be seen with  `srepr(P(Y>3))`. When you print `P(Y>3)`, the str printer produces a more readable form of the expression. In this process it moves its bits around. So when you then copy-paste the result, the new expression is not exactly like the original. It's annoying. You can raise it as an issue. As for the test, something like
```
P(Y>3).equals(32*(-S(71)/32 + 3*exp(4)/32)*exp(-4)/3)
```
would be sufficient.

smichr replied:
Try `((-S(71)/32 + 3*exp(4)/32)*exp(-4))/3*32` instead.

normalhuman commented on sympy/stats/tests/test_discrete_rv.py:
```diff
@@ -34,3 +36,12 @@ def test_sample():
+    X = Geometric('X', S(1)/5)
+    Y = Poisson('Y', 4)
+    assert P(Eq(X, 3)) == S(16)/125
+    assert P(X < 3) == S(9)/25
```
I suggest also adding tests for <= and >= relations.

normalhuman commented on sympy/stats/tests/test_discrete_rv.py:
```diff
@@ -34,3 +36,12 @@ def test_sample():
+    assert P(Eq(X, 3)) == S(16)/125
+    assert P(X < 3) == S(9)/25
+    assert P(X > 3) == S(64)/125
+    assert P(Eq(Y, 3)) == 32*exp(-4)/3
```
I also tried `P(Ne(Y, 3))` and got
```
AttributeError: 'SingleDiscretePSpace' object has no attribute 'random_symbol'
```

akash9712 committed:
Modified method `probability` in SingleDiscretePSpace

Modified the method `probability` to evaluate the result in situations where
the domain is a union of various sets.
Added and modified tests.

akash9712 committed:
Resolved merge conflicts.

akash9712 commented:
@normalhuman I have made the suggested changes. Please review.

normalhuman reviewed (commented):
What's with conditional probability, for example
```
X = Geometric("X", S(7)/9)
P(X>9, X>3)
```
throws "AttributeError: 'SingleDiscretePSpace' object has no attribute 'conditional_space'".

It does the same in master, but since you are looking at these, maybe it can be fixed too? If not, I'll put it up as a separate issue.

asmeurer commented on sympy/stats/drv.py:
```diff
@@ -157,3 +164,44 @@ def compute_characteristic_function(self, expr, **kwargs):
+
+    def probability(self, condition):
+        _domain = self.restricted_domain(condition)
+        if condition is False or _domain is S.EmptySet:
```
You should use `== False` instead of `is False`. If a symbolic condition like `Eq(x, x)` evaluates to a boolean, it will evaluate to a SymPy Boolean (`S.true` or `S.false`), not a Python bool.

asmeurer replied:
Also these cases are not tested.

asmeurer commented on sympy/stats/drv.py:
```diff
@@ -157,3 +164,44 @@ def compute_characteristic_function(self, expr, **kwargs):
+            return S.One
+        try:
+            return self.eval_prob(_domain)
+        except:
```
Do not use bare except. This will mask any error in eval_prob. Is there any reason eval_prob cannot raise NotImplementedError on its own? This also should check for a None return type and raise NotImplementedError then.

asmeurer replied:
Actually shouldn't it just return an unevaluated object?

akash9712 commented:
@normalhuman @asmeurer Classes `FinitePSpace` and `ContinuousPSpace` are defined for finite and continuous random variables, whereas there is no such class for discrete random variables. As far as I understand, these classes allow the creation of random variables with any given domain and density. Should a similar class be created for the discrete case as well? That should solve the issue with conditional probability as well.

akash9712 committed:
Changes to eval_prob in sympy.stats.drv

eval_prob raises error in case probability for the given condition
is not implemented. An unevaluated `Probability` object is then returned.

normalhuman commented:
There are also FiniteDomain, FiniteDensity, ProductFiniteDomain, ProductFinitePSpace,... with their analogues for continuous distributions. Some of these may need discrete analogues. Perhaps this should be a separate project though. SymPy has too many PRs that tried to do too much and were never merged.

akash9712 commented:
@normalhuman +1
I'll update the PR with the changes required by the review, and will try to send a seperate PR that implements some of these classes.

akash9712 committed:
Added tests for discrete RV `probability`

Added tests to check for boundary conditions.

normalhuman commented:
Looks good to me.

akash9712 commented:
@normalhuman Regarding the classes like DiscreteDomain to be implemented, would it be a good idea to represent the domains using an Imageset? I am asking because the `sup` or `inf` for an ImageSet is not currently implemented. Or should the current implementation be restricted to Range (which would not allow not integer step sizes for the discrete domain)?

normalhuman commented:
drv.py effectively assumes the step of 1 already ([example](https://github.com/sympy/sympy/blob/master/sympy/stats/drv.py#L105)); without this assumption nothing works. Given this, and the fact nobody came up with non-integer-step infinite discrete distributions worth modeling, I would go with Range.

Upabjojr commented on sympy/stats/drv.py:
```diff
@@ -157,3 +165,45 @@ def compute_characteristic_function(self, expr, **kwargs):
+            rv = sum(self.eval_prob(x) for x in _domain.args)
+            return rv
+        else:
+            raise NotImplementedError(filldedent('''Probability for
```
we actually have a class for unevaluated probability, it's called `Probability`.

akash9712 replied:
The class is being used on L189. The unevaluated object is returned if this function raises a `NotImplemented` error at this point. Since the exception raised at this point is always caught, the error message is never actually displayed.

Upabjojr replied:
Oh, I see. I would recommend anyway to try to avoid using exception handling, as it makes it difficult to debug the code. Every time you raise and catch an exception you lose the stack trace. Also the readability of the code is worsened by raising exceptions.

akash9712 replied:
Thanks for explaining. Should I remove the part where `eval_prob` raises the exception, and modify `probability` so that it returns the unevaluated object in case the return value from `eval_prob` is `None`?

akash9712 replied:
I have opened #14604 for this, including one other change.
